Initial review of APE 10

astrofrog opened the pull request.

astrofrog committed:
Started APE 10 from template

astrofrog committed:
Initial version of APE 10

astrofrog committed:
Minor wording

astrofrog committed:
Added a section about deviations from APE 2

astrofrog committed:
Fix syntax

astrofrog committed:
Wording

astrofrog commented:
@eteq @taldcroft - here's a very early draft of an APE re: python 2/3 support. Could you take a look over this and let me know what you think? Once you are happy with, I can open a proper PR to the main repository and we can send it around for comments?

astrofrog commented on APE10.rst:
```diff
@@ -0,0 +1,144 @@
+  is because PyPI does not compare the current Python version to the PyPI
+  meta-data for the package (which may indicate for example whether the package
+  is Python 2-compatible), and will download Astropy v3.0 regardless of the
+  active Python version. However, there is a solution, which is …
```
There is a solution that involves inserting a suffix in the PyPI tarball name. I've asked @Carreau if he can provide me with links that will provide a reference for this.

Carreau replied:
https://github.com/jupyter/roadmap/blob/master/accepted/migration-to-python-3-only.md#multiple-source-distributions

```
Multiple source distributions

Alternatively, we could use a hidden feature of pip. Source distributions whose base name ends in -pyX.Y are considered for installation only on a matching Python version. Thus instead of publishing a single .tar.gz, we could publish multiple identical source distribution with various names.

For supported python versions:

ipython-6.x-py3.4.tar.gz
ipython-6.x-py3.5.tar.gz
For future python versions:

ipython-6.x-py3.6.tar.gz
ipython-6.x-py3.7.tar.gz
It may be possible to patch pip to handle a more general form like ipython-6.x-py3.tar.gz, so we only need to publish a single sdist.
```

Which is triggerd by https://github.com/pypa/pip/blob/a942fb9331762e88992c51fc3cf2ddffd5e4f3fb/pip/index.py#L635-L643

replacing `py_version != sys.version[:3]:` by `py_version != sys.version[:len(py_version)]:` should be enough. 

PyPI also support extra metadata [see https://pypi.python.org/pypi/flit] which says for example `Requires Python: 3` on PyPI UI, though it can't be set through setuptool, and is not supported by pip.

astrofrog replied:
@Carreau - thanks for all the information! Will fold this in to this document.

astrofrog committed:
Expanded section on Python 3-only releases on PyPI

astrofrog committed:
Fix link

astrofrog commented:
pip discussion in https://github.com/pypa/pip/pull/3847

taldcroft commented on APE10.rst:
```diff
@@ -0,0 +1,155 @@
+A number of scientific Python projects, including IPython, Matplotlib, Sympy,
+and others have now signed the `Python 3 statement
+<https://python3statement.github.io>`_ which states that scientific Python
+packages will drop support for Python 2 no longer than 2020. In fact, the
```
`later` not `longer`

eteq replied:
Also on this line: maybe "In fact" -> "already" or something like that? ("in fact" implies sort of surprise, and we want to make this _less_ surprising)

taldcroft commented on APE10.rst:
```diff
@@ -0,0 +1,155 @@
+and others have now signed the `Python 3 statement
+<https://python3statement.github.io>`_ which states that scientific Python
+packages will drop support for Python 2 no longer than 2020. In fact, the
+IPython developers have now released `IPython 5.0
```
Even though IPython is clearly an important part of the ecosystem, I'm not sure this level of detail for their support plan is warranted.  I think it would be more important to compare with numpy, scipy, and matplotlib as key dependencies.

eteq replied:
Yeah, I think it's fine to just say that the 5.x is the last that will suport py 2.x.  Although you _could_ move this below and say it's motivation for _our_ roadmap.

taldcroft commented on APE10.rst:
```diff
@@ -0,0 +1,155 @@
+In practice, dropping Python 2.7 support will involve, amongst other things:
+
+* Removing the bundled `six <https://pythonhosted.org/six/>`_ package, and
+  update all the code that used ``astropy.extern.six`` to use the Python 3
```
`updating`

taldcroft commented on APE10.rst:
```diff
@@ -0,0 +1,155 @@
+* Removing any backports that may exist for Python 2.7 compatibility
+* Updating the documentation to remove any mention of Python 2.7, as well as
+  making sure that Python 3.x is not presented as a special case.
+* Remove Python 2 and 2.7 from the PyPI classifiers
```
`Removing`

taldcroft commented on APE10.rst:
```diff
@@ -0,0 +1,155 @@
+  who will then be able to help their colleagues also make the transition.
+
+* We will be able to start using Python 3-only features internally, including
+  for example function annotations (e.g., for units), matrix multiplication
```
Doesn't matrix multiplication need Python 3.5?  Do we (or should we) have an official policy on number of Python versions supported?  I know this has been discussed previously and there tend to be different opinions.

astrofrog replied:
Very good question, and actually I think this belongs in this APE too. I think we should follow the normal Python support dates. Python 3.4 will no longer be supported from early 2019 so we could switch to using `@` at that point (we're currently not supporting 3.1 and 3.2 which aren't supported by Python either).

eteq replied:
That makes sense.  Definitely should _say_ that here, though, that we intend to generally follow the Python support timeline?

taldcroft commented on APE10.rst:
```diff
@@ -0,0 +1,155 @@
+  making sure that Python 3.x is not presented as a special case.
+* Remove Python 2 and 2.7 from the PyPI classifiers
+
+Challenges
```
Perhaps another challenge and point to address is the actual PR or set of PRs to do the 2 to 3 migration after 2.0 comes out.  It seems like it would be good to have a moratorium on other PRs being merged during this process and should probably be the first thing to happen after 2.0 is released.  

Right now there are 130 open PRs.  I imagine a fair number of PRs will get merge conflicts due to the 2 -> 3 migration, though maybe it won't be too horrible.

taldcroft replied:
Another challenge: maintaining the 2.0 LTS will be more difficult than for 1.0 because development bug fixes will not be guaranteed to be Python-2 compatible.

astrofrog replied:
Good point - I'll add both of these!

astrofrog committed:
Use tables for release calendars

astrofrog committed:
Implemented suggestions by @taldcroft

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,178 @@
+following:
+
+==========   ==================  ==========================================
+Version      Release date        Notes
```
`Release date` -> `(Planned) Release date`?  b/c some are already done and some are future.  Maybe also mark which ones have happened as of this APE just so that no one looking at this in the future gets concerned if it seems the dates here don't quite match the reality of future releases.

astrofrog replied:
Done - I actually split the table in two to avoid confusion.

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,178 @@
+Note that v3.0 would not be an LTS release, since we would
+otherwise need to maintain two LTS releases in parallel. This is a one-off
+deviation from the version numbering policy outlined in `APE 2
+<https://github.com/astropy/astropy-APEs/blob/master/APE2.rst>`_.
```
And maybe add something like: "and subsequent releases will follow the pattern established by APE 2.

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,178 @@
+
+The purpose of this APE is to set out a roadmap for future releases of the
+Astropy core package, in particular to identify the process for switching to
+releases that support only Python 3.
```
Maybe also add here also the specific fact of "ending support for py 2.x in December 2019"

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,178 @@
+* Removing any backports that may exist for Python 2.7 compatibility
+* Updating the documentation to remove any mention of Python 2.7, as well as
+  making sure that Python 3.x is not presented as a special case.
+* Removing Python 2 and 2.7 from the PyPI classifiers
```
I'm not sure about this one: shouldn't it still be there because _past_ versions will still support it?  Or are the classifiers distinct for each _version_ as opposed to the whole package?

astrofrog replied:
I'm pretty sure the classifiers/metadata on PyPI are version-specific.

eteq commented:
Overall looking good, @astrofrog !

One thing that I _think_ is missing (although maybe you have it in there and I missed that), is the fact that the Python core devs will be dropping support for 2.x - that's probably important motivation for doing this in the first place.

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,178 @@
+  for example function annotations (e.g., for units), matrix multiplication
+  (e.g., for coordinates; note that this will only be possible once we support
+  only Python 3.5+)
+
```
One more advantage to add (one of the biggest in my mind!): It will make maintaining and writing the code significantly easier, because devs don't have to know both 2.x and 3.x (and `six`) and the various "gotchas" involved there.  They can just code consistently for py 3.x only.

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,178 @@
+
+This would allow more time for the PyPI limitations mentioned above to be
+resolved and more time for developers to make the transition to Python 3. On the
+other hand, if we actually want users to virtually all be using Python 3 by
```
remove "virtually", since this is talking about a "want", anyway.

Carreau commented:
> Another challenge: maintaining the 2.0 LTS will be more difficult than for 1.0 because development bug fixes will not be guaranteed to be Python-2 compatible.

Or you fix the bug on LTS, and forward-port.

astrofrog committed:
Implemented comments by @eteq and @taldcroft

astrofrog commented:
@taldcroft @eteq - I've now implemented all your comments - thanks for the initial review! Do you want me to include you as authors or would it make more sense for you to sit this one out in case coordination committee decisions are needed?

@Carreau - thanks for your comment too, I've added a sentence to this effect. Feel free to read over the document and leave any comments if you are interested!

Note that if anyone wants to read over the rendered version, you can see it [here](https://github.com/astrofrog/astropy-APEs/blob/ape-10/APE10.rst)

taldcroft commented:
@astrofrog - I don't need to be an author.

taldcroft commented:
Looks good to me now.

astrofrog committed:
Add note to clarify what ending support means

eteq commented on APE10.rst:
```diff
@@ -0,0 +1,214 @@
+v3.1         June 2018
+v3.2         December 2018
+v3.3         June 2019
+v4.0 - LTS   December 2019            Supported with bug fixes for two years)
```
Looks like there's a trailing `)` here that's not not supposed to be there?

eteq commented:
One minor typo I noted above, but otherwise looks good to me!

astrofrog committed:
Minor typo fix

astrofrog committed:
Updated date

astrofrog committed:
Small wording improvements for pip

astrofrog committed:
Updated table
